src/game/ingame_menu.c

```diff
@@ -224,15 +210,184 @@ void create_dl_ortho_matrix(void) {
+    struct Utf8CharLUTEntry *usedLUT;
+    u32 length;
+
+    lut = segmented_to_virtual(lut);
+    if (!(str[*strPos] & 0x20)) {
+        codepoint = ((str[*strPos] & 0x1F) << 6) | (str[*strPos + 1] & 0x3F);
+        *strPos += 1;
+
+        usedLUT = segmented_to_virtual(lut->lut2Bytes);
+        length = lut->length2Bytes;
+    } else if (!(str[*strPos] & 0x10)) {
+        codepoint = ((str[*strPos] & 0xF) << 12) | ((str[*strPos + 1] & 0x3F) << 6) | (str[*strPos + 2] & 0x3F);
+        *strPos += 2;
+
+        usedLUT = segmented_to_virtual(lut->lut3Bytes);
+        length = lut->length3Bytes;
+    } else {
+        codepoint = ((str[*strPos] & 0x7) << 18) | ((str[*strPos + 1] & 0x3F) << 12) | ((str[*strPos + 2] & 0x3F) << 6) | (str[*strPos + 3] & 0x3F);
+        *strPos += 3;
+
+        usedLUT = segmented_to_virtual(lut->lut4Bytes);
+        length = lut->length4Bytes;
+    }
+
+    s32 start = 0;
+    s32 end = length - 1;
+    s32 mid = (start + end) / 2;
+
+    while (start <= end) {
+        if (usedLUT[mid].codepoint == codepoint) {
+            return &usedLUT[mid];
+        }
+
+        if (usedLUT[mid].codepoint > codepoint) {
+            end = mid - 1;
+        } else {
+            start = mid + 1;
+        }
+
+        mid = (start + end) / 2;
+    }
+    return segmented_to_virtual(lut->missingChar);
+}
+
+/**
+ * Convert a character in the range 0-9 or A-F to an integer value. Returns -1 if the character is
+ * not a valid hex digit.
+ */
+static s32 hex_char_to_value(char c) {
+    if (c >= '0' && c <= '9') {
+        return c - '0';
+    }
+    if (c >= 'A' && c <= 'F') {
+        return c - 'A' + 10;
+    }
+    if (c >= 'a' && c <= 'f') {
+        return c - 'a' + 10;
+    }
+    return -1;
+}
+
+/**
+ * Determine if the characters following an @ sign are a valid hex color code.
+ */
+static s32 is_color_code_valid(char *str, s32 strPos) {
+    for (s32 i = 0; i < 6; i++) {
+        if (hex_char_to_value(str[strPos + i]) == -1) {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+/**
+ * Get the exact width of the line of a string of any font in pixels, using the given ASCII and UTF-8 tables.
+ */
+s32 get_string_width(char *str, struct AsciiCharLUTEntry *asciiLut, struct Utf8LUT *utf8LUT) {
+    s32 width = 0;
+    s32 maxWidth = 0;
+    s32 strPos = 0;
+    char c;
+
+    s32 isGenericFont = (asciiLut == main_font_lut);
+
+    asciiLut = segmented_to_virtual(asciiLut);
+
+    while ((c = str[strPos]) != '\0' && c != '\n') {
+        // Handle color codes and tabs if using generic font
+        if (isGenericFont) {
+            if (c == CHAR_COLOR_CODE) {
+                if (is_color_code_valid(str, strPos + 1)) {
+                    strPos += 7;
+                    continue;
+                }
+            } else if (c == '\t') {
+                width += 4 * SPACE_KERNING(asciiLut);
+                strPos++;
+                continue;
+            }
+        }
+        if (c & 0x80) {
```

Comment:
is this to check if its multibyte?
```suggestion
                if (is_color_code_valid(str, strPos + 1)) {
                    strPos += 7;
                    continue;
                }
            } else if (c == '\t') {
                width += 4 * SPACE_KERNING(asciiLut);
                strPos++;
                continue;
            }
        }
        
        // check if multibyte sequence
        if (c & 0x80) {
```